daemon/emer-daemon.c

```diff
@@ -1123,6 +1142,379 @@ set_max_bytes_buffered (EmerDaemon *self,
+                              event_id_str);
+
+  return g_build_filename (priv->persistent_cache_directory,
+                           "aggregate-timers",
+                           date_string,
+                           filename,
+                           NULL);
+}
+
+static guint32
+find_previous_aggregate_timer_counter (AggregateTimerData *timer_data)
+{
+  g_autofree gchar *tally_path = NULL;
+  g_autofree gchar *contents = NULL;
+  g_autoptr(GVariant) variant = NULL;
+  g_autoptr(GVariant) cache_key = NULL;
+  g_autoptr(GVariant) event_id = NULL;
+  g_autoptr(GVariant) aggregate_key = NULL;
+  g_autoptr(GVariant) payload = NULL;
+  g_autoptr(GError) error = NULL;
+  gboolean complete;
+  guint32 previous_counter;
+  uid_t unix_user_id;
+
+  tally_path = get_tally_path_from_aggregate_timer (timer_data);
+
+  if (!g_file_test (tally_path, G_FILE_TEST_EXISTS))
+    return 0;
+
+  g_file_get_contents (tally_path, &contents, NULL, &error);
+  if (error)
+    return 0;
+
+  variant = g_variant_parse (NULL, contents, NULL, NULL, &error);
+  if (error)
+    return 0;
+
+  g_variant_get (variant,
+                 "(vvvuubmv)",
+                 &event_id,
+                 &aggregate_key,
+                 &cache_key,
+                 &unix_user_id,
+                 &previous_counter,
+                 &complete,
+                 &payload);
+
+  g_assert (complete);
+
+  return previous_counter;
+}
+
+static void
+save_aggregate_timer_to_tally (AggregateTimerData *timer_data)
+{
+  g_autofree gchar *printed_variant = NULL;
+  g_autofree gchar *tally_path = NULL;
+  g_autoptr(GVariant) timer_variant = NULL;
+  g_autoptr(GError) error = NULL;
+  guint32 previous_counter;
+
+  /* It might happen that the same application was opened and closed
+   * multiple times at the same day by the same user. In this case,
+   * we want to increase the day's counter instead of simply override
+   * it.
+   */
+  previous_counter =
+    find_previous_aggregate_timer_counter (timer_data);
+
+  timer_variant = aggregate_timer_to_variant (timer_data,
+                                              previous_counter);
+  printed_variant = g_variant_print (timer_variant, TRUE);
+
+  tally_path = get_tally_path_from_aggregate_timer (timer_data);
+
+  g_file_set_contents (tally_path,
+                       printed_variant,
+                       -1,
+                       &error);
+
+  if (error)
+    {
+      g_warning ("Error saving aggregate timer to tally: %s",
+                 error->message);
+    }
+}
+
+static void
+save_aggregate_timers_to_tally (EmerDaemon *self)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  AggregateTimerData *timer_data;
+  GHashTableIter iter;
+
+  g_hash_table_iter_init (&iter, priv->aggregate_timers);
+  while (g_hash_table_iter_next (&iter, NULL, (gpointer *) &timer_data))
+    {
+      save_aggregate_timer_to_tally (timer_data);
+      timer_data->start_monotonic_us = g_get_monotonic_time ();
```

Comment:
Suppose there are 100 running timers. We will call g_get_monotonic_time() 200 times here – once in aggregate_timer_to_variant (via save_aggregate_timer_to_tally) and again here, for each event. Of course it is likely not to have any significant impact on the time in seconds but it would be preferable IMO to use the same monotonic time when cutting the timer in two, as the end of the first part and the start of the second part, and having done that we could use the same monotonic time for all of them.